daemon/attach.go

```diff
@@ -113,138 +113,109 @@ func (daemon *Daemon) ContainerAttach(job *engine.Job) engine.Status {
-				log.Errorf("attach: stdout pipe: %s", err)
+		wg.Add(1)
+		log.Debugf("attach: stdin: begin")
+		defer func() {
+			if stdinOnce && !tty {
+				cStdin.Close()
 			} else {
-				io.Copy(&ioutils.NopWriter{}, cStdout)
+				// No matter what, when stdin is closed (io.Copy unblock), close stdout and stderr
+				if cStdout != nil {
+					cStdout.Close()
+				}
+				if cStderr != nil {
+					cStderr.Close()
+				}
 			}
+			wg.Done()
+			log.Debugf("attach: stdin: end")
 		}()
-	}
-	if stderr != nil {
-		nJobs++
-		if p, err := streamConfig.StderrPipe(); err != nil {
-			errors <- err
+		var err error
+		if tty {
+			_, err = utils.CopyEscapable(cStdin, stdin)
 		} else {
-			cStderr = p
-			go func() {
-				log.Debugf("attach: stderr: begin")
-				defer log.Debugf("attach: stderr: end")
-				// If we are in StdinOnce mode, then close stdin
-				// Why are we closing stdin here and above while handling stdout?
-				if stdinOnce && stdin != nil {
-					defer stdin.Close()
-				}
-				_, err := io.Copy(stderr, cStderr)
-				if err == io.ErrClosedPipe {
-					err = nil
-				}
-				if err != nil {
-					log.Errorf("attach: stderr: %s", err)
+			_, err = io.Copy(cStdin, stdin)
+
+		}
+		if err == io.ErrClosedPipe {
+			err = nil
+		}
+		if err != nil {
+			log.Errorf("attach: stdin: %s", err)
+			return err
+		}
+		return nil
+	})
+
+	attachStream := func(name string, stream io.Writer, streamPipe io.ReadCloser) chan error {
+		if stream == nil {
+			chErr := make(chan error, 1)
+			chErr <- nil
+			return chErr
+		}
+		return promise.Go(func() error {
+			wg.Add(1)
+			defer func() {
+				// Make sure stdin gets closed
+				if stdinOnce && cStdin != nil {
+					stdin.Close()
+					cStdin.Close()
 				}
-				errors <- err
+				streamPipe.Close()
+				wg.Done()
+				log.Debugf("attach: %s: end", name)
 			}()
-		}
-	} else {
-		// Point stderr at a no-op writer.
-		go func() {
-			if cStderr, err := streamConfig.StderrPipe(); err != nil {
-				log.Errorf("attach: stdout pipe: %s", err)
-			} else {
-				io.Copy(&ioutils.NopWriter{}, cStderr)
-			}
-		}()
-	}
 
-	return promise.Go(func() error {
-		defer func() {
-			if cStdout != nil {
-				cStdout.Close()
-			}
-			if cStderr != nil {
-				cStderr.Close()
+			log.Debugf("attach: %s: begin", name)
+			_, err := io.Copy(stream, streamPipe)
+			if err == io.ErrClosedPipe {
+				err = nil
 			}
-		}()
-
-		for i := 0; i < nJobs; i++ {
-			log.Debugf("attach: waiting for job %d/%d", i+1, nJobs)
-			if err := <-errors; err != nil {
-				log.Errorf("attach: job %d returned error %s, aborting all jobs", i+1, err)
+			if err != nil {
+				log.Errorf("attach: %s: %v", name, err)
 				return err
 			}
-			log.Debugf("attach: job %d completed successfully", i+1)
+			return nil
+		})
+	}
+
+	stdoutErr = attachStream("stdout", stdout, cStdout)
+	stderrErr = attachStream("stderr", stderr, cStderr)
+
+	return promise.Go(func() error {
+		wg.Wait()
+		if err := <-stdinErr; err != nil {
```

Comment:
Okay, I was wrong about first err :) But I still don't like three channels and three `promise.Go`. Also I hate `promise` :cat: